crates/turborepo-scm/src/ls_tree.rs

```diff
@@ -0,0 +1,143 @@
+use std::{
+    io::{BufRead, BufReader, Read},
+    process::{Command, Stdio},
+};
+
+use nom::Finish;
+use turbopath::{AbsoluteSystemPathBuf, RelativeUnixPathBuf};
+
+use crate::{package_deps::GitHashes, read_git_error, wait_for_success, Error};
+
+pub fn git_ls_tree(root_path: &AbsoluteSystemPathBuf) -> Result<GitHashes, Error> {
+    let mut hashes = GitHashes::new();
+    let mut git = Command::new("git")
+        .args(["ls-tree", "-r", "-z", "HEAD"])
+        .current_dir(root_path)
+        .stdout(Stdio::piped())
+        .stderr(Stdio::piped())
+        .spawn()?;
+
+    let stdout = git
+        .stdout
+        .as_mut()
+        .ok_or_else(|| Error::git_error("failed to get stdout for git ls-tree"))?;
+    let mut stderr = git
+        .stderr
+        .take()
+        .ok_or_else(|| Error::git_error("failed to get stderr for git ls-tree"))?;
+    let result = read_ls_tree(stdout, &mut hashes);
+    if let Err(err) = result {
+        return Err(read_git_error(&mut stderr).unwrap_or(err));
+    }
+    wait_for_success(git, &mut stderr, "git ls-tree", root_path)?;
+    Ok(hashes)
+}
+
+fn read_ls_tree<R: Read>(reader: R, hashes: &mut GitHashes) -> Result<(), Error> {
+    let mut reader = BufReader::new(reader);
+    let mut buffer = Vec::new();
+    loop {
+        buffer.clear();
+        {
+            let bytes_read = reader.read_until(b'\0', &mut buffer)?;
+            if bytes_read == 0 {
+                break;
+            }
+            {
+                let entry = parse_ls_tree(&buffer)?;
+                let hash = String::from_utf8(entry.hash.to_vec())?;
+                let path = RelativeUnixPathBuf::new(entry.filename)?;
+                hashes.insert(path, hash);
+            }
+        }
+    }
+    Ok(())
+}
+
+struct LsTreeEntry<'a> {
+    filename: &'a [u8],
+    hash: &'a [u8],
+}
+
+fn parse_ls_tree(i: &[u8]) -> Result<LsTreeEntry<'_>, Error> {
+    match nom_parse_ls_tree(i).finish() {
```

Comment:
Do we care if there's text left over? Generally with combinators we want to ensure that there's nothing left

Reply:
Added some `all_consuming` calls.